packages/react-reconciler/src/ReactFiberClassComponent.js

```diff
@@ -513,27 +513,52 @@ function constructClassInstance(
   let unmaskedContext = emptyContextObject;
   let context = null;
   const contextType = ctor.contextType;
-  if (typeof contextType === 'object' && contextType !== null) {
-    if (__DEV__) {
-      const isContextConsumer =
+
+  if (__DEV__) {
+    if ('contextType' in ctor) {
+      let isValid =
+        contextType !== null &&
+        contextType !== undefined &&
         contextType.$$typeof === REACT_CONTEXT_TYPE &&
-        contextType._context !== undefined;
-      if (
-        (contextType.$$typeof !== REACT_CONTEXT_TYPE || isContextConsumer) &&
-        !didWarnAboutInvalidateContextType.has(ctor)
-      ) {
+        contextType._context === undefined; // Not a <Context.Consumer>
+
+      if (!isValid && !didWarnAboutInvalidateContextType.has(ctor)) {
         didWarnAboutInvalidateContextType.add(ctor);
+
+        let addendum = '';
+        if (contextType === null) {
+          addendum = ' However, it is set to null.';
+        } else if (contextType === undefined) {
+          addendum =
+            ' However, it is set to undefined. ' +
+            'This can be caused by a typo or by mixing up named and default imports. ' +
+            'This can also happen due to a circular dependency, so ' +
+            'try moving the createContext() call to a separate file.';
+        } else if (typeof contextType !== 'object') {
+          addendum = ' However, it is set to ' + typeof contextType + '.';
+        } else if (contextType.$$typeof === REACT_PROVIDER_TYPE) {
+          addendum = ' Did you accidentally pass the Context.Provider instead?';
+        } else if (contextType._context !== undefined) {
+          // <Context.Consumer>
+          addendum = ' Did you accidentally pass the Context.Consumer instead?';
+        } else {
```

Comment:
Add another branch for arrays?

```js
else if (Array.isArray(contextType)) {
  addendum = 'However, it is set to an array.'
}
```

Reply:
Seems very unlikely to happen

Reply:
🤷‍♂️ seems like it could happen if you import the wrong value. I figured that if it happens it would be weird to see.

> However, it is set to an object with keys {0, 1, 2, 3, 4, 5, 6...100000}

Reply:
Arrays are pretty rare as export values so I don't think it's worth handling.